### DISCUSSION TITLE
Building issues

### DISCUSSION THREAD

**Discussion by CopperEagle** [jmltoolkit/jml-tools#3] [Q&A]
Hey there @wadoon 

Thanks for the repository. For my project, I need to translate jml to java code (into assertions). Unfortunately there is no documentation.  I wanted to ask you

1) How do I build this into a jar usable from Java?

2) How can I use your jml2java module from Java?

I tried

```
# gradle init
gradle build
```
but I end up getting a few-byte jar file both with and without init. Also, the Github action build keeps failing too so that's not helping either.

Any help is appreciated

**Comment by wadoon** [comment]
You can use the regular Gradle targets, e.g., `gradle classes`, `testClasses`, or `assemble` to compile the class files, and also the test classes, or to finally build all artifacts. I also fixed some compiler errors.

The functionalities are rather in an unfinished state. For example you mentioned `jml2java` module: Some expressions are supported, but support for complicated constructs (especially, `\old`) is absent. 

Currently, we are migrating [JJBMC](https://github.com/wadoon/JJBMC/tree/jonasVersion) from openjml to jmlparser, which also implements a jml2java conversion feature, but designed for bounded model checker (e.g., use of nondet. variables). 

/cc @JonasKlamroth

**Reply by CopperEagle** [reply]
Hello @wadoon 

Thank you very much for your response. I greatly appreciate your help with this matter. 

My current priority is to utilize JML for specifying class invariants. So I need a tool that can effectively translate these invariants into assertion statements. In your opinion, would JJBMC be a suitable choice for this purpose? Additionally, I would be interested to know if there are any alternative projects or tools that you are aware of that offer similar capabilities for converting JML to Java. I value your insights and recommendations on this topic. 

Once again, thank you for your assistance. 

Regards

**Reply by JonasKlamroth** [reply]
Hey @CopperEagle ,
your work sounds very interesting. JJBMC is a tool which does a translation from JML to Java in order to run a bounded model checker on that translated code. Thus the translation is currently not available as a library or jar that you could easily use. However if you have problems translating certain JML-constructs you could have a look at it. 

Regarding other tools: The only thing that comes to my mind regarding this is using the RAC (runtime assertion checking) mode of OpenJML. This sounds pretty close to what you would want to do but I'm not an expert for OpenJML. However, as far es i know this feature of OpenJML is also not fully supported and either deprecated or not yet fully implemented. So I'm not sure if it is even better than this repo.

**Reply by wadoon** [reply]
It is possible to synthesize a method for most class invariants, that returns to true iff the invariant is adhered to in the current state. 

However, some JML constructs are incomputable, e.g., consider Hilbert operator `(\some int x; ....)`, or unbounded quantification. Other constructs are semantically unclear, e.g., `(\forall Object o; o.hashCode() > 0)` (in KeY this quantifies over every possible object, in OpenJML it means the current reachable objects). 

A different underdefined point is the states/locations where the invariant needs to hold. There are currently two to three semantics for invariants floating around. 

`jmlparser` and `jml-tools` give you the tools to parse, transform and modify JML/Java source code, but they are intentionally semantics-free. The semantics is something you need to bring in.

**Reply by CopperEagle** [reply]
Thank you very much for your insights @wadoon and @JonasKlamroth. I will have a look at the RAC and thank you for pointing out the different semantical nuances concerning invariants!

Have a nice day
Kind regards

**Reply by CopperEagle** [reply]
Dear @wadoon and @JonasKlamroth 

Regarding `invariant`s I have a question. I was suprised to find out that the official JML reference by David R. Cok, Gary T. Leavens, and Mattias Ulbrich does not specify what the visibility modifiers mean, e.g. `//@ `*`private`*` invariant head == 0`. How does the JML community typically interpret this?

Am I right to guess that it modifies for which methods the invariant applies, as follows:
- `public invariant`: Only for public methods
- `protected invariant`: Only *public and* protected ones
- package invariant (no modifier): Only package and public ones
- `private invariant`: Any method

Is this right? Or is it typically interpreted more restrictive?
Once again, thank you very much for your assistance.

Kind regards

**Reply by wadoon** [reply]
The modifier interpretation is only for visibility reasons. 

The reason is that an invariant can express facts on private, protected or internal variables (or methods), that might not be visible to the caller/user of a method/class. 

Therefore, like in Java, using private fields in public invariants is a bad choice. 

Now it depends on your reasoning and the scope of your invariant utilization. Hence, how do you use the invariants in your tooling? 

In general, all invariants need to hold on method calls. But with an external checker, you might only use public invariants, as the access is limited to those members.

KeY completely ignores all visibility modifiers. 

If you want to disable the class invariants on a method, the method must have the `helper` modifier (at least in KeY). Invariants of object parameters (e.g., `foo(o)`) are not assumed automatically in KeY.

**Reply by CopperEagle** [reply]
Dear @wadoon 

Thank you for your response! In the tool I write, the invariants have access to all fields and are only for a specific class. It is then used to test this class. So there is no passing of specifications up or down the class hierarchy.

I was assuming that there is some mechanism in JML invariants itself to steer where they apply. Private methods typically are private because they violate some invariant and thus should not be used from the outside. However, the approach using the `helper` modifier is indeed more finegrained and makes more sense. 

Also, I wanted to ask whether you could recommend some real projects/classes that use JML (so I can test my tool on them). The only industrial example I know so far that uses JML is the encryption-sdk by [AWS](https://github.com/aws/aws-encryption-sdk-java). Could you point out some other datastructure implementations or projects that make use of JML?

Thank you very much for your answer!

**Reply by wadoon** [reply]
I have gathered most of the case studies in the KeY organization: 

https://github.com/keyproject/

There are: 

* https://github.com/KeYProject/ips4o-verify
  New very fast sorting algorithm. TACAS'24


* https://github.com/KeYProject/VerifyingIdentityHashMap
  java.util.IdentityHashMap -- Best paper, iFM'22

* https://github.com/KeYProject/BitSet-KeY-Artifact
  java.util.BitSet, FACS'23

* https://github.com/KeYProject/rbtree-verification
  Red-Black Trees

* https://github.com/KeYProject/DualPivotQuickSort
  Dual Pivot Quicksort: One of the standard sorting algorithms in Java. VSTTE'17?

* https://github.com/KeYProject/TimSort
  TimSort -- also a JDK default sorting algorithm. CAV'15. Bug found. 

* https://github.com/KeYProject/verifythis-ltc-2020
  Simplified Version of the Hagrid key-server

**Reply by CopperEagle** [reply]
@wadoon Oh, I did not yet know KeY had some examples on their Github. Thank you so much for your support! It helped a lot!
